### PR TITLE
nn layer functions 

### PR DESCRIPTION
**Pull request by Maximellerbach**
quite some changes, need to write some new tests + backprop

### CODE REVIEW

**Review by mickael-btc — APPROVED** [review]
nice !

**Comment by Maximellerbach** [comment]
still need to fully implement and debug conv forward & backprop.
There are some things I need to debug also on the derivative of cross entropy -> making gradient explode / being nan. 

We can merge this in a first time, then I will fix and rework what is needed to be
